event type setting refactor

Udit-takkar opened the pull request:
Fixes https://github.com/calcom/cal.com/issues/5946



**Environment**: Staging(main branch) / Production

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Chore (refactoring code, technical debt, workflow improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

Commit:
wip

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
refactor: fix types and logic

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
chore: remove comment

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

hariombalhara commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -401,7 +401,14 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered
-                    const inputIndex = customInputs.findIndex((input) => input.id === values.id);
+                    const inputIndex = customInputs.findIndex(
+                      (input) =>
+                        input.id === values.id &&
```
This seems hackish. id should uniquely identify an element, why do we need to match all props

Udit-takkar replied:
so id is only assigned after user has clicked save button otherwise a default id is being used -1. so editing multiple custom inputs without saving would work something like this https://photos.google.com/share/AF1QipMCbabiovDIRlru0ASfB0uerBr-aoN4XLG0_xTwAerAzAm1FFeRA4Kp-yf-Yi4bWA/photo/AF1QipP42jgzOHOeraQ45gF0MFcHCRxcEhjfl1dfyMq1?key=ZkZ0ZnZhQ2d0UXNYdE8wSHFmaWJjNTdsaWdGRmp3

It would overwrite other inputs which have same id.

hariombalhara replied:
Can we generate id(from uuid()) at frontend ? That should solve the problem.

Udit-takkar replied:
so i tried generating random id ( can't use uuid() because we want id of type number. i tried something else here).  but a lot of logic in https://github.com/calcom/cal.com/blob/main/packages/trpc/server/routers/viewer/eventTypes.tsx#L38 depends on id being <0 ( -1 which is default).

Udit-takkar replied:
@hariombalhara  okay i was facing some bugs earlier now i am using Math.random to generate id because type required is number only.

added a new optional variable which is being used to detect if custom input has to be created or updated .Earlier we were using id<0 (-1 which is default) to check new custom inputs https://github.com/calcom/cal.com/blob/main/packages/trpc/server/routers/viewer/eventTypes.tsx#L38 

and lastly useEffect to update customInput after saving.

this should work now

hariombalhara commented on apps/web/pages/event-types/[type]/index.tsx:
```diff
@@ -99,33 +92,38 @@ const querySchema = z.object({
-
-  const router = useRouter();
-  const { tabName } = querySchema.parse(router.query);
+  const { data, isLoading } = trpc.viewer.eventTypes.get.useQuery({ id: props.type });
```
It would be better to move useQuery to a wrapper Component for `EventTypePage` component. So, loading logic would be limited to the wrapper component and thus EventTypePage would have non null eventType. 
This would avoid all the newly added `eventType?` and `eventType&&` things

hariombalhara replied:
We should use initialData with useQuery that is obtained from SSR. That makes the first page load really fast with no loader and later navigations to eventType page would be client side rendered.

Udit-takkar replied:
Done

hariombalhara reviewed (changes requested):
Great attempt to improve EventType flow @Udit-takkar !! 
Just need a bit of work

Commit:
fix: minor changes

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: use wrpper component

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: remove optional chaing

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: use initial date in useQuery

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Udit-takkar commented:
@hariombalhara  thanks for the review. made all the changes you suggested.

Commit:
fix: type errors

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

hariombalhara commented on apps/web/components/eventtype/EventTeamTab.tsx:
```diff
@@ -1,5 +1,5 @@
 import { SchedulingType } from "@prisma/client/";
-import { EventTypeSetupInfered, FormValues } from "pages/event-types/[type]";
+import { EventTypeSetupProps, FormValues } from "pages/event-types/[type]";
```
Let's clean these unused import as well.

hariombalhara commented on packages/trpc/server/routers/viewer/eventTypes.tsx:
```diff
@@ -434,15 +440,233 @@ export const eventTypesRouter = router({
       }
-      return await ctx.prisma.eventType.findUnique({
+
+      const userSelect = Prisma.validator<Prisma.UserSelect>()({
```
This logic is duplicate of getServerSideProps, let's reuse it. Maybe we can create a getEventType fn or something similar that is used by both getServerSideProps and tRPC.

The only difference I see is how 404 and other errors are handled.

hariombalhara replied:
@Udit-takkar Can you take care of this as well.

Udit-takkar replied:
@hariombalhara done

hariombalhara reviewed (commented):
Looks better now. Left some more comments.

Commit:
chore: remove unused import

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: type error

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: use only id to update custom inputs

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

PeerRich commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -386,20 +392,22 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered
                 onSubmit={(values) => {
                   const customInput: CustomInputParsed = {
-                    id: -1,
+                    id: Math.floor(Math.random() * 10000),
```
whats going on here? 🤣

Udit-takkar replied:
we can't use uuid of type string so had to generate id of type number. number of custom inputs won't be too many so there should not be any collision.

Btw this is just temporary id till user click save. real id would be assigned after user click save and custom inputs would re render with correct id

zomars replied:
Can we keep the number negative? That way we can GUARANTEE avoiding collision

Udit-takkar replied:
It would cause this bug if we use id as -1
https://photos.google.com/share/AF1QipMCbabiovDIRlru0ASfB0uerBr-aoN4XLG0_xTwAerAzAm1FFeRA4Kp-yf-Yi4bWA/photo/AF1QipP42jgzOHOeraQ45gF0MFcHCRxcEhjfl1dfyMq1?key=ZkZ0ZnZhQ2d0UXNYdE8wSHFmaWJjNTdsaWdGRmp3 .

bug is also mentioned here https://github.com/calcom/cal.com/issues/5946. the likelihood of collision is very less as after saving we would get the correct id from server. if you want i can try to further improve the id generation  Math.floor(Math.random() * 10000).

Udit-takkar replied:
@zomars ok added negative

PeerRich commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -386,20 +392,22 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered
                   const customInput: CustomInputParsed = {
-                    id: -1,
+                    id: Math.floor(Math.random() * 10000),
                     eventTypeId: -1,
```
what is this -1?

Udit-takkar replied:
-1 was used to denote new custom input which was used here to detect whether we have to create a field or update it https://github.com/calcom/cal.com/blob/main/packages/trpc/server/routers/viewer/eventTypes.tsx#L38. i just added another property in the customInputParsed type to denote if we have to create a new input or update it.

Udit-takkar replied:
-1 as id was causing bugs when you create more than one custom input and try to edit them without clicking the save button on top.

zomars replied:
https://github.com/calcom/cal.com/pull/5959#discussion_r1052525718

hariombalhara replied:
`-1 * Math.random()` could still give the same negative number for 2 different custom inputs.
The simplest solution that I can think of is to use `Date.now()` as 2 different custom inputs are manually added there would always be a time difference and thus 2 different `Date.now `call will have guaranteed different values.

Udit-takkar replied:
@hariombalhara date.now() was the first thing i tried. there was some prisma error because of the size.

Udit-takkar replied:
we can only send 32 bit signed integer as id just updated the random id generator function to this. This function would give the temporary id till the user clicks save and then we would get the correct id from the server which won't have any collision. 

```
const getRandomId = (length = 8) => {
  return (
    -1 *
    parseInt(
      Math.ceil(Math.random() * Date.now())
        .toPrecision(length)
        .toString()
        .replace(".", "")
    )
  );
};
```

Commit:
fix: use negative random id

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: improve random id generation

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: create common function for getting event id

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

alannnc commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -51,7 +63,7 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered
   const [requiresConfirmation, setRequiresConfirmation] = useState(eventType.requiresConfirmation);
   const placeholderHashedLink = `${CAL_URL}/d/${hashedUrl}/${eventType.slug}`;
-
+  console.log(customInputs);
```
```suggestion

```

alannnc committed:
Update apps/web/components/eventtype/EventAdvancedTab.tsx

alannnc commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -386,20 +403,22 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered
                 onSubmit={(values) => {
                   const customInput: CustomInputParsed = {
-                    id: -1,
+                    id: getRandomId(),
```
Can we use uuid() generators instead of random number?

Udit-takkar replied:
No because the type of id is number and not string

Commit:
chore: fix types

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

alannnc reviewed (commented):
Can we do loom? Or self review? I'm having trouble to understand why some changes are needed.

Udit-takkar commented:
@alannnc anything specific do you wanna know about?

Udit-takkar commented on packages/lib/getEventTypeById.ts:
```diff
@@ -0,0 +1,264 @@
+import { Prisma, PrismaClient } from "@prisma/client";
```
this is just a common function which is used in getServersideprops and trpc end point. refer here https://github.com/calcom/cal.com/pull/5959#discussion_r1050806695 on why we are using it twice.

Udit-takkar commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -64,6 +75,12 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered
 
+  useEffect(() => {
+    if (eventType.customInputs) {
+      setCustomInputs(eventType.customInputs.sort((a, b) => a.id - b.id));
```
this makes sure the custom inputs are re rendered when we click save button.

Udit-takkar commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -36,7 +36,19 @@ const generateHashedLink = (id: number) => {
 };
 
-export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered, "eventType" | "team">) => {
+const getRandomId = (length = 8) => {
```
customInput type requires the id to be of type number and within 32 bit integer so i created this random id generator function. This is the temporary id we'll be using until user clicks save and server save these custom inputs and return correct id.  

our current approach of using id as -1 was causing this bug https://photos.google.com/share/AF1QipMCbabiovDIRlru0ASfB0uerBr-aoN4XLG0_xTwAerAzAm1FFeRA4Kp-yf-Yi4bWA/photo/AF1QipP42jgzOHOeraQ45gF0MFcHCRxcEhjfl1dfyMq1?key=ZkZ0ZnZhQ2d0UXNYdE8wSHFmaWJjNTdsaWdGRmp3

Udit-takkar commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -386,20 +403,22 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupInfered
                     required: values.required,
                     type: values.type,
                     options: values.options,
+                    hasToBeCreated: true,
```
this variable would be used in handleCustomInputs(https://github.com/calcom/cal.com/blob/main/packages/trpc/server/routers/viewer/eventTypes.tsx#L38) to detect if we have to create a custom input. we can also choose to not make this change this but this makes it more obvious that we are creating custom input.

Udit-takkar commented on apps/web/components/eventtype/EventTeamTab.tsx:
```diff
@@ -101,48 +100,3 @@ export const EventTeamTab = ({
   );
 };
-
-const TeamProfileCard = ({
```
A lot of stuff was not used anywhere. so i removed it

Udit-takkar commented on apps/web/pages/event-types/[type]/index.tsx:
```diff
@@ -109,34 +105,34 @@ const querySchema = z.object({
   });
 
-  const { eventType: dbEventType, locationOptions, team, teamMembers } = props;
-  // TODO: It isn't a good idea to maintain state using setEventType. If we want to connect the SSR'd data to tRPC, we should useQuery(["viewer.eventTypes.get"]) with initialData
```
i basically did this stuff which was also mentioned in TODO.

Udit-takkar commented on apps/web/pages/event-types/[type]/index.tsx:
```diff
@@ -326,6 +315,18 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
   );
 };
 
+const EventTypePageWrapper = (props: inferSSRProps<typeof getServerSideProps>) => {
```
this wrapper makes sure EventTypePage is loaded only when data is not null or undefined else we would have to use too much of optional chaining and && operator. ( suggested by  hariom)

Udit-takkar commented on packages/trpc/server/routers/viewer/eventTypes.tsx:
```diff
@@ -36,30 +37,29 @@ function handlePeriodType(periodType: string | undefined): PeriodType | undefine
 
 function handleCustomInputs(customInputs: CustomInputSchema[], eventTypeId: number) {
-  const cInputsIdsToDelete = customInputs.filter((input) => input.id > 0).map((e) => e.id);
+  const cInputsIdsToDeleteOrUpdated = customInputs.filter((input) => !input.hasToBeCreated);
```
we were filtering same thing twice so i just created another variable

Udit-takkar reviewed (commented):
Self review done. @alannnc  do let me know if i missed something

hariombalhara commented:
<img width="1177" alt="Screenshot 2022-12-22 at 1 15 53 PM" src="https://user-images.githubusercontent.com/1780212/209083522-9ca40055-b551-4f91-87cb-6e961ea50863.png">
I see hydration errors in console for event-types/[type] page

zomars committed:
Link fixes

commit 56c1608deaf1921a80dc71bcbd107e772dd656f4
Author: Omar López <zomars@me.com>
Date:   Wed Dec 21 19:26:04 2022 -0700

    Update link.ts

commit eab43d833515cabc7acf2f973c2d2837ac1a0332
Author: zomars <zomars@me.com>
Date:   Wed Dec 21 19:03:02 2022 -0700

    Prevents 500 on TRPC error

commit 3a6970d5e5fab54136b379e1b2ccd7acf515a809
Author: zomars <zomars@me.com>
Date:   Wed Dec 21 18:15:26 2022 -0700

    Encoding fixes

zomars committed:
Lowercase usename checks again.

Udit-takkar commented:
https://user-images.githubusercontent.com/53316345/209204464-d9ed4699-1f0d-4fdc-9c3f-f861d306ae15.mov

@hariombalhara  I don't get any hydration error

hariombalhara commented:
@Udit-takkar can you try refreshing the page

Udit-takkar commented:
@hariombalhara yeah on refreshing i get hydration error but this error is also on the main branch. what are possible fix for this?

hariombalhara commented:
You would need to identify the difference b/w  server rendered content and client rendered content. That difference is the cause for hydration error. 
Error message should help. 

Though if it's in main, it isn't a blocker. But as  we are improving event-type page, you can tackle it in this PR itself

Commit:
fix: conflicts

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: add workflows

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Udit-takkar committed:
Merge branch 'main' into fix/event-type-refactor

Commit:
chore: remove unused

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

PeerRich committed:
Merge branch 'main' into fix/event-type-refactor

hariombalhara reviewed (approved):
Approving it now. The only issue that I see now is hydration error which is already in production and can be fixed separately 

Awesome work @Udit-takkar 🙏

kodiakhq[bot] committed:
Merge branch 'main' into fix/event-type-refactor
